### PR TITLE
feat: Add experimental CLI for LoopBack 4

### PR DESCRIPTION
**Pull request by raymondfeng**
### Description

Usage:
```sh
npm i -g @loopback/cli
lb4 app <name>
lb4 extension <name>
```

For reviewers, you can check out the branch and run `node packages/cli/bin/cli app <my-app>`  or `node packages/cli/bin/cli extension <my-ext>`.

#### Related issues

- connect to https://github.com/strongloop/loopback-next/issues/361

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

File: packages/cli/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+```
+Usage:
+  lb4 app [options] [<name>]
+
+Options:
+  -h,   --help             # Print the generator's options and usage
+        --skip-cache       # Do not remember prompt answers              Default: false
+        --skip-install     # Do not automatically install dependencies   Default: false
+        --applicationName  # Application name
+        --description      # Description for theapplication
```
Comment (b-admike) [inline packages/cli/README.md]: `the application`

Comment (raymondfeng) [reply packages/cli/README.md]: Good catch. Fixed.

File: packages/cli/test/project.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+              ],
+            },
+            props
+          )
+        );
+      });
+
+      it('creates files', () => {
+        assert.noFile(['tslint.json', 'tslint.build.json']);
+        assert.jsonFileContent('package.json', props);
```
Comment (b-admike) [inline packages/cli/test/project.js]: I wonder how `assert.jsonFileContent` works. Below we are checking for `prettier` [here](https://github.com/strongloop/loopback-next/pull/693/files#diff-626d225f72681fb6a0d8254766ba46e6R151), but not in this particular test?

**Comment by b-admike** [comment]
Trail run result (for app):
```
| => node packages/cli/bin/cli app
? Project name: foobar
? Project description: my first lbnext app!
? Project root directory: foobar
? Application class name: FoobarApplication
? Select project build settings:  Enable tslint, Enable prettier, Enable mocha, Enable loopb
ackBuild
   create .gitignore
   create .npmrc
   create .prettierignore
   create .prettierrc
   create index.d.ts
   create index.js
   create index.ts
   create LICENSE
   create package.json
   create README.md
   create src/application.ts
   create src/controllers/ping-controller.ts
   create src/controllers/README.md
   create src/index.ts
   create src/repositories/README.md
   create test/mocha.opts
   create test/ping-controller.test.ts
   create test/README.md
   create tsconfig.json
   create tslint.build.json
   create tslint.json

___________________    | ~/loopback/loopback-next @ biniams-mbp (badmike)
| => rm -rf foobar/
___________________    | ~/loopback/loopback-next @ biniams-mbp (badmike)
| => node packages/cli/bin/cli app
? Project name: foobar
? Project description: my first lbnext app!
? Project root directory: foobar
? Application class name: FoobarApplication
? Select project build settings:  Enable tslint, Enable prettier, Enable mocha, Enable loopb
ackBuild
   create .gitignore
   create .npmrc
   create .prettierignore
   create .prettierrc
   create index.d.ts
   create index.js
   create index.ts
   create LICENSE
   create package.json
   create README.md
   create src/application.ts
   create src/controllers/ping-controller.ts
   create src/controllers/README.md
   create src/index.ts
   create src/repositories/README.md
   create test/mocha.opts
   create test/ping-controller.test.ts
   create test/README.md
   create tsconfig.json
   create tslint.build.json
   create tslint.json

> foobar@1.0.0 prepare /Users/badmike/loopback/loopback-next/foobar
> npm run build


> foobar@1.0.0 build /Users/badmike/loopback/loopback-next/foobar
> lb-tsc

added 570 packages in 31.853s

Application foobar is now created in foobar.

Next steps:

$ cd foobar
$ npm start

___________________    | ~/loopback/loopback-next @ biniams-mbp (badmike)
| => cd foobar/
___________________    | ~/loopback/loopback-next/foobar @ biniams-mbp (badmike)
| => npm start

> foobar@1.0.0 start /Users/badmike/loopback/loopback-next/foobar
> npm run build && node .


> foobar@1.0.0 build /Users/badmike/loopback/loopback-next/foobar
> lb-tsc

Server is running at http://127.0.0.1:3000
Try http://127.0.0.1:3000/ping
Unhandled error in GET /favicon.ico: 404 Error: Endpoint "GET /favicon.ico" not found.
    at RoutingTable.find (/Users/badmike/loopback/loopback-next/foobar/node_modules/@loopback/rest/dist/src/router/routing-table.js:75:15)
    at HttpHandler.findRoute (/Users/badmike/loopback/loopback-next/foobar/node_modules/@loopback/rest/dist/src/http-handler.js:26:29)
    at DefaultSequence.request [as findRoute] (/Users/badmike/loopback/loopback-next/foobar/node_modules/@loopback/rest/dist/src/providers/find-route.js:29:40)
    at DefaultSequence.handle (/Users/badmike/loopback/loopback-next/foobar/node_modules/@loopback/rest/dist/src/sequence.js:83:32)
    at HttpHandler._handleRequest (/Users/badmike/loopback/loopback-next/foobar/node_modules/@loopback/rest/dist/src/http-handler.js:32:24)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
```

`ping` page:
```
{"greeting":"Hello from LoopBack","date":"2017-11-02T23:19:24.416Z","url":"/ping","headers":{"host":"localhost:3000","connection":"keep-alive","user-agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36","upgrade-insecure-requests":"1","accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8","accept-encoding":"gzip, deflate, br","accept-language":"en-US,en;q=0.8"}}
```

File: packages/cli/generators/app/templates/src/controllers/ping-controller.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  constructor(@inject('rest.http.request') private req: ServerRequest) {}
+
+  // Map to `GET /ping`
+  @get('/ping')
+  ping(): object {
+    return {
+      greeting: 'Hello from LoopBack',
+      date: new Date(),
+      url: this.req.url,
+      headers: Object.assign({}, this.req.headers),
```
Comment (b-admike) [inline packages/cli/generators/app/templates/src/controllers/ping-controller.ts]: I found this to be a bit ugly for UX, but feel free to keep it.

File: packages/cli/generators/app/templates/src/repositories/README.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Repositories
```
Comment (b-admike) [inline packages/cli/generators/app/templates/src/repositories/README.md]: The `Mixin`, `Providers`, etc.. pages have a great deal of description. Can we add some links to our docs or brief explanations here as well?

File: packages/cli/generators/extension/templates/src/controllers/README.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Controllers
```
Comment (b-admike) [inline packages/cli/generators/extension/templates/src/controllers/README.md]: The `Mixin`, `Providers`, etc.. pages have a great deal of description. Can we add some links to our docs or brief explanations here as well?

File: packages/cli/generators/extension/templates/src/repositories/README.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# Repositories
```
Comment (b-admike) [inline packages/cli/generators/extension/templates/src/repositories/README.md]: The `Mixin`, `Providers`, etc.. pages have a great deal of description. Can we add some links to our docs or brief explanations here as well?

File: packages/cli/generators/extension/templates/test/acceptance/README.md [hunk 7]
```diff
@@ -0,0 +1 @@
+# Acceptance tests
```
Comment (b-admike) [inline packages/cli/generators/extension/templates/test/acceptance/README.md]: The `Mixin`, `Providers`, etc.. pages have a great deal of description. Can we add some links to our docs or brief explanations here as well?

File: packages/cli/generators/extension/templates/test/integration/README.md [hunk 8]
```diff
@@ -0,0 +1 @@
+# Integration tests
```
Comment (b-admike) [inline packages/cli/generators/extension/templates/test/integration/README.md]: The `Mixin`, `Providers`, etc.. pages have a great deal of description. Can we add some links to our docs or brief explanations here as well?

File: packages/cli/generators/extension/templates/test/unit/README.md [hunk 9]
```diff
@@ -0,0 +1 @@
+# Unit tests
```
Comment (b-admike) [inline packages/cli/generators/extension/templates/test/unit/README.md]: The `Mixin`, `Providers`, etc.. pages have a great deal of description. Can we add some links to our docs or brief explanations here as well?

**Comment by raymondfeng** [comment]
> Unhandled error in GET /favicon.ico: 404 Error: Endpoint "GET /favicon.ico" not found.

This is expected as the browser will send `GET /favicon.ico` to the server and it's not available. In LoopBack 3.x, we ship a default LoopBack logo as the icon. The runtime should have not dump 404 errors by default, IMO. It should be configured and toggled somehow.

File: packages/cli/generators/app/templates/src/repositories/README.md (hunk 4 above)
Comment (raymondfeng) [reply packages/cli/generators/app/templates/src/repositories/README.md]: The READMEs are copied from https://github.com/strongloop/loopback4-extension-starter. @virkt25 Do you make the docs for Mixin/Provider/Decorator available to http://loopback.io/doc/en/lb4/Concepts.html? We should only keep links in the README.md to be scaffolded into user projects.

File: packages/cli/generators/app/templates/src/controllers/ping-controller.ts (hunk 3 above)
Comment (raymondfeng) [reply packages/cli/generators/app/templates/src/controllers/ping-controller.ts]: Haha, I have a json viewer extension installed for my browser and it looks fine :-). Surprisingly, most browsers don't render json response nicely by default.

File: packages/cli/test/project.js (hunk 2 above)
Comment (raymondfeng) [reply packages/cli/test/project.js]: See https://github.com/yeoman/yeoman-assert#api. The assertions here focus on verifying the omission of tslint related artifacts.

**Comment by bajtos** [comment]
Consider our past experience with yeoman, I think we should not be basing any new tooling on yeoman. We should use [inquirer](https://www.npmjs.com/package/inquirer) and [ejs templating](https://www.npmjs.com/package/ejs) directly.

Having said that, I am happy to use yeoman for this initial experimental implementation, as it makes it easier for us to get something up and running 👍 

I don't have bandwidth to review this pull request in details, I'll leave that up to @strongloop/sq-lb-apex

**Comment by raymondfeng** [comment]
> Considering our past experience with yeoman, I think we should not be basing any new tooling on yeoman. We should use inquirer and ejs templating directly.

Do we have a list of pain points with Yeoman somewhere? The latest version seems to be working out pretty well. I was able to register generators with yeoman-environment explicitly without the magic and the module name doesn't have to be `generator-*`.

**Comment by kjdelisle** [comment]
I'm not well-versed enough in templating and generators to have an informed opinion. I'd probably have to take a few days to play around with it and see what I think. If we're concerned about blocking this PR for that long, then I won't stand in the way.

**Comment by raymondfeng** [comment]
@kjdelisle I would like to have it merged as soon as possible so that we can start to use it to scaffold new extensions/apps. Maybe we can move forward as follows:

1. I'll merge the PR and make a new release
2. Some of us can start to use it after `npm i -g @loopback/cli`.
3. We'll refine the templates as they are being improved over time.

**Comment by jonathan-casarrubias** [comment]
When can we expect this experimental cli to be published?

**Comment by jonathan-casarrubias** [comment]
Ok cool, installing....

**Pull request merged**